### PR TITLE
Feature/file upload

### PR DESCRIPTION
**Pull request by CarlHembrough**
Initial commit of full implementation.

The solution hosts a webpage on http://localhost:20019/ allowing a file to be uploaded. When the file is uploaded its sent to S3 for storage and a event is posted to Kafka.

### CODE REVIEW

File: file/s3/store.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager/s3manageriface"
+	"io"
+)
+
+// NewFileStore factory method to initialise AWS S3 classes.
+func NewFileStore(AWSRegion string, s3Bucket string) *FileStore {
+	return &FileStore{
+		Uploader:   s3manager.NewUploader(session.New(&aws.Config{Region: aws.String(AWSRegion)})),
```
Comment (NeilMadden) [inline file/s3/store.go]: Can any errors/panics happen here?

File: handler/upload.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+var FailedToReadRequest string = "Failed to read upload file from the request."
+var FailedToSaveFile string = "Failed to save the given file."
+var FailedToSendEvent string = "Failed to send file uploaded event."
+
+func Upload(w http.ResponseWriter, req *http.Request) {
+
+	if FileStore == nil {
+		log.Error(errors.New("The FileStore dependency has not been configured"), log.Data{})
```
Comment (NeilMadden) [inline handler/upload.go]: Can you pass `nil` instead of an empty `log.Data{}` here?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	router := pat.New()
+	alice := alice.New(
+		timeout.Handler(10*time.Second),
+		log.Handler,
+		requestID.Handler(16),
+	).Then(router)
+
+	router.Get("/healthcheck", healthcheck.Handler)
+	router.Get("/", handlers.Home)
+	router.Post("/", handlers.Upload)
```
Comment (NeilMadden) [inline main.go]: Do we (eventually) need any access control or rate limiting on this upload feature?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	log.Namespace = "dp-dd-file-uploader"
+
+	bindAddr := os.Getenv("BIND_ADDR")
+	if len(bindAddr) == 0 {
+		bindAddr = config.BindAddr
+	}
+
+	var err error
+	render.Renderer = unrolled.New()
+	handlers.FileStore = s3.NewFileStore(config.AWSRegion, config.S3Bucket)
```
Comment (NeilMadden) [inline main.go]: Presumably the S3 SDK picks up credentials from the environment?

**Comment by NeilMadden** [comment]
I don't think you meant to commit the .DS_Store

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Change log
+==========
+
+0.2.0 (2016-10-24)
+- Another example changelog entry
```
Comment (CarlHembrough) [inline CHANGELOG.md]: Done

File: file/s3/store.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	}
+}
+
+// FileStore S3 implementation
+type FileStore struct {
+	Uploader   s3manageriface.UploaderAPI
+	BucketName string
+}
+
+// SaveFile ...
```
Comment (CarlHembrough) [inline file/s3/store.go]: Done

File: file/s3/store.go (hunk 1 above)
Comment (CarlHembrough) [inline file/s3/store.go]: There are no errors returned to handle.

File: handler/upload.go (hunk 2 above)
Comment (CarlHembrough) [inline handler/upload.go]: Yes - I have made the change throughout.

File: main.go (hunk 3 above)
Comment (CarlHembrough) [inline main.go]: AFAIK access control will be provided via florence eventually. Not sure about the rate limiting but its a good question.

File: main.go (hunk 4 above)
Comment (CarlHembrough) [inline main.go]: Yes, either from environment variables or known credential file location ( ~/.aws )

File: LICENSE.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+The MIT License (MIT)
```
Comment (githubmo) [inline LICENSE.md]: Why aren't we using the government open source license? 

https://gds.blog.gov.uk/2012/10/12/coding-in-the-open/

https://en.wikipedia.org/wiki/Open_Government_Licence

**Pull request merged**